Mark a deadline task as finished

nightism opened the pull request:
Deadline task now can be marked as finished by command `fin-deadline INDEX`.
All the finished deadline task will be hidden from the user interface, but not removed from the storage.

Commit:
Task: isFinished attribute added in DeadlineTask

Commit:
TaskTest: update DeadlineTaskTest

Commit:
TaskBook: add "mark deadline as finished" function

Commit:
TaskBookTest: update TaskBookTest for markDeadlineFinished method

Commit:
storage: update JsonDeadlineTaskMixin for new attribute "isFinished"

Commit:
storage: update JsonFloatingTaskMixinTest

Commit:
Model: add "mark deadline as finished" function support

Commit:
ModelTest: update ModelTest for markDeadlineFinished method

Commit:
Command: implement MarkDeadlineFinishedCommand

Commit:
parser: add MarkDeadlineFinishedCommand support

nightism commented:
This is a required function in V0.3, please merge it if there is no major problem.
Revision could be done in the future.

nightism commented:
"mark finished" for floating task will be followed up as soon as possible.

nightism commented:
However, for event task, I think there is not need to have "mark finished" function.

pyokagan commented on src/main/java/seedu/address/model/Model.java:
```diff
@@ -109,4 +112,11 @@
+     * The current filter should be guaranteed not to be a null.
+     */
+    void refilterDeadlineTaskFilteredList();
+
```
Ugh, shouldn't `setDeadlinetaskFilter` be sufficient?

nightism replied:
Set the same filter will not render the list be refiltered again.

nightism replied:
But I also do not think the current method to refilter the list is a good one.

pyokagan commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -241,4 +255,12 @@ public void setDeadlineTaskFilter(Predicate<? super DeadlineTask> predicate) {
+        Predicate<? super DeadlineTask> predicate = filteredDeadlineTasks.getPredicate();
+        assert predicate != null;
+        setDeadlineTaskFilter(null);
+        setDeadlineTaskFilter(predicate);
```
There's no need to set it to `null` first.

pyokagan replied:
And this shows that actually `setDeadlineTaskFilter()` is sufficient and there's no need to introduce a new thing to the interface just for it.

pyokagan replied:
Ah.. no wonder.. If I'm guessing correctly it's because the UI is not updating? (Because you did not obey the "Tasks are immutable" rule)

nightism replied:
> Ah.. no wonder.. If I'm guessing correctly it's because the UI is not updating? (Because you did not obey the "Tasks are immutable" rule)

Yes, you are correct.
But if we just create a new task every time we want to finish it, this function will be just the same as `edit`.

pyokagan replied:
> But if we just create a new task every time we want to finish it, this function will be just the same as edit.

What's wrong with that?

I mean, if you think about it, _technically_ it is the same. Both are editing the fields of a task.

pyokagan commented on src/main/java/seedu/address/model/task/DeadlineTask.java:
```diff
@@ -3,28 +3,50 @@
+    public void markAsFinished() {
+        assert this.finished == false;
+        this.finished = true;
+    }
```
Ugh, **nope**. Tasks **must** be immutable. If you allow tasks to be mutated, you are going to get lots of problems with (e.g., the UI, task book saving on change, etc.)

pyokagan replied:
To add on: it will also make @potatowagon 's life harder because we won't be able to just do a shallow copy anymore to save the state the taskbook for undo/redo (The task instance will keep its previous values)

nightism replied:
Emmm... so I just create a new task every time we want to mark it as finish?

pyokagan replied:
> Emmm... so I just create a new task every time we want to mark it as finish?

Yes.

If you think that requires too much code, consider using the [builder pattern](https://immutables.github.io/immutable.html#builder)

pyokagan commented on src/main/java/seedu/address/model/task/DeadlineTask.java:
```diff
@@ -3,28 +3,50 @@
+    public void markAsFinished() {
+        assert this.finished == false;
+        this.finished = true;
+    }
```
In fact, I guess that with your current implementation, if you mark a deadline task as finished, the task book will not be saved, which is a bug ;-)

pyokagan reviewed (changes requested):
Tasks must be immutable. If this rule is not obeyed, there are going to be lots of subtle bugs in the application.

pyokagan commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -232,7 +232,21 @@ public synchronized void setDeadlineTask(int indexInFilteredList, DeadlineTask n
+    public synchronized DeadlineTask markDeadlineFinished(int indexInFilteredList)
+            throws IllegalValueException {
+        final DeadlineTask finishedDeadline = taskBook.markDeadlineFinished(getDeadlineTaskSourceIndex(indexInFilteredList));
+        refilterDeadlineTaskFilteredList();
```
So later, when we implement the "find tasks by keywords" command, marking a deadline as completed will reset the filter? Would that be what the user expects?

nightism replied:
emmm... yes, that's a problem...

Commit:
Model: delete markDeadLineFinished method

* refilteredDeadlineTaskFilteredList also removed from model

Commit:
TaskBook: remove markDeadlineFinished method

Commit:
Task: DeadlineTask implementation revised

* markAsFinished removed to prevent mutable data
* add new constructor which takes in an oddTask and a new finishedStatus

Commit:
Command: MarkDeadlineFinishedCommand.execute() revised

Instead of revising the attribute of the old task, this command will create a new task to replace the old one.

nightism commented:
@pyokagan new implementation has been done, please have a look

pyokagan commented on src/main/java/seedu/address/model/task/DeadlineTask.java:
```diff
@@ -3,28 +3,49 @@
+    }
+
+    public DeadlineTask(DeadlineTask oldTask, boolean finishStatus) {
+        this(oldTask.name, oldTask.getDue(), finishStatus);
```
Ugh, I don't think these kind of constructors should be introduced, because then for consistency (and consistency is important for good code) we will have to introduce constructors for every combination of fields that we potentially want to edit. That is, we need to implement 7 constructors!

Either just use `new DeadlineTask(oldName, oldDue, newCompleted)`, or use the builder pattern

pyokagan commented on src/main/java/seedu/address/model/task/DeadlineTask.java:
```diff
@@ -3,28 +3,49 @@
     private final LocalDateTime due;
 
-    public DeadlineTask(Name name, LocalDateTime due) {
+    private boolean finished;
```
This can be `final` now.

pyokagan commented on src/main/java/seedu/address/model/task/DeadlineTask.java:
```diff
@@ -3,28 +3,49 @@
+
+    public static Predicate<? super DeadlineTask> isNotFinishedDeadline() {
+        return p -> !p.isFinished();
+    }
```
Is there any reason why this has to be a part of `DeadlineTask`? Don't we already have the `seedu.address.model.filter` package for that putting all the model filters inside?

pyokagan commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -232,7 +232,12 @@ public synchronized void setDeadlineTask(int indexInFilteredList, DeadlineTask n
+     */
     public ObservableList<DeadlineTask> getFilteredDeadlineTaskList() {
+        // filter out all finished deadline task
+        setDeadlineTaskFilter(DeadlineTask.isNotFinishedDeadline());
```
**Nope**. Getter methods must not cause any change in state.

pyokagan replied:
I think we should just leave out filtering for this round. Modifying the model to have a `completed` field is a good first step, and filtering can be done on top.

pyokagan reviewed (changes requested):
`getFilteredDeadlineTaskList()` must **not** cause any changes in state.

By the way, a completed task is not the same as a deleted task. Shouldn't there be some indication in the UI as to whether the task is completed or not?

pyokagan commented:
@nightism Didn't I already review this iteration?

Commit:
revised all points mentioned in pr conversation

Commit:
Filter: implement DeadlineTaskFinished

* to be used in future implementation

Commit:
ui: update ui to display the `finished` status of deadline task

Commit:
TaskTest: updated with DeadlineTaskTest

pyokagan commented on src/main/java/seedu/address/logic/commands/MarkDeadlineFinishedCommand.java:
```diff
@@ -0,0 +1,48 @@
+
+public class MarkDeadlineFinishedCommand extends Command {
+
+    public static final String COMMAND_WORD = "fin-deadline";
```
Probably should be `fin-due` to be consistent with the rest of the commands, but eh, they should be replaced in the future anyway.

pyokagan commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -232,7 +232,11 @@ public synchronized void setDeadlineTask(int indexInFilteredList, DeadlineTask n
     @Override
+    /**
+     * Return the filtered deadline task list, all finished deadline task will be filtered out
+     */
```
hmm, checkstyle didn't complain about this? Usually the javadoc should be before the annotation.

pyokagan replied:
Oh I just realised this is an unneeded change :-/

pyokagan reviewed (approved):
Looks good, thanks. I'll fix up the one minor style error and merge.

pyokagan commented on src/main/java/seedu/address/model/ModelManager.java:
```diff
@@ -196,7 +196,7 @@ public void setEventTaskFilter(Predicate<? super EventTask> predicate) {
     public synchronized void addDeadlineTask(DeadlineTask deadlineTask) {
         taskBook.addDeadlineTask(deadlineTask);
-        setDeadlineTaskFilter(null);
+        assert deadlineTask.isFinished() == false;
```
Ugh, I just noticed that this code has actually been here since the start...

Hmmmm..... We will have to rethink this when we start implementing the filtering commands.

pyokagan replied:
So actually, this change should not remove the `setDeadlineTaskFilter`

pyokagan commented:
Merged as b61184cfdbadb57242b55c490824ce365122bf49
